integrate vm-memory with dirty page tracking

alindima opened the pull request:
# Reason for This PR

The vm-memory crate from rust-vmm has recently introduced abstractions for dirty page tracking, similar to the ones we had in Firecracker.
We need to consume them to avoid code duplication and to strip down the proxy crate we have in Firecracker.

Note that this PR is made on a feature branch.

## Description of Changes

- remove the Firecracker code responsible for dirty page tracking and the bitmap abstractions.
- add a single function for creating the `GuestMemoryMmap` object, `vm_memory::create_guest_memory`. There were a lot of alternatives of creating the guest memory before this PR, which were hard to follow.
- place and refactor the existing code for guard pages in the `vm_memory::create_guest_memory` function.
- replace the deprecated usage of `.with_regions()` and `.with_regions_mut()` with `.iter()`.
- add a test utility for creating unguarded guest memory. This is needed because some tests need memory regions with sizes that are not multiples of 4096 (the default page size).
- replace all test instantiations of `GuestMemoryMmap` with calls to `create_guest_memory` or the `create_guest_memory_unguarded` test utility (where needed).
- port the unit tests for guard pages

The proxy crate is still needed to reexport the GuestMemory* types with the concrete type of `B: Option<AtomicBitmap>`. Otherwise we'd have to modify all the code that uses the GuestMemory* types, which wouldn't look nice.
It's also used for exporting the `create_guest_memory()` function.

- [ ] This functionality can be added in [`rust-vmm`](https://github.com/rust-vmm).

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The reason for this PR is clearly provided (issue no. or explanation).
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] Any newly added `unsafe` code is properly documented.
- [x] Any API changes are reflected in `firecracker/swagger.yaml`.
- [x] Any user-facing changes are mentioned in `CHANGELOG.md`.
- [x] All added/changed functionality is tested.

acatangiu commented:
Seems like upstream feature is not fully stabilized yet: https://github.com/rust-vmm/vm-memory/pull/158

alindima commented:
> Seems like upstream feature is not stabilized yet: [rust-vmm/vm-memory#158](https://github.com/rust-vmm/vm-memory/pull/158)

Thanks for pointing it out, that PR seems very fresh. I'm going to make this PR a draft and keep it this way until rust-vmm has a release

alindima commented:
vm-memory 0.6.0 was released so I'll rebase this PR soon and mark it ready for review

alindima commented:
PR is now ready for review

serban300 commented on src/vm-memory/src/lib.rs:
```diff
@@ -155,3 +155,205 @@ pub mod test_utils {
     }
 }
+
+#[cfg(test)]
```
Can we move this commit right after the first one ? Or in the first one. Since it seems to add tests for the methods that are defined there.

serban300 reviewed (commented):
The PR looks good. I just have a couple of small comments, most of them related to the old code. As a high level comment, we usually make sure that each commit builds. I don't know of any exception so far. Can we do the same thing here while keeping the PR structure (keeping it split in multiple small commits in order to make it easy to review) ?

sandreim commented on src/vm-memory/src/lib.rs:
```diff
@@ -5,18 +5,355 @@
+}
+
+/// Helper for creating the guest memory.
+pub fn create_guest_memory(
```
This vm-memory crate was purposed as a proxy crate (to allow specialization or adding functionality missing in upstream atm) until the feature gap closes. Dirty page tracking in upstream brings use one step closer to dropping this crate from FC.

This PR this is challenging that by further expanding on functionality/scope.  Do you think we should revisit that decision or strip it as much as possible in this PR ?

alindima replied:
> This PR this is challenging that by further expanding on functionality/scope. Do you think we should revisit that decision or strip it as much as possible in this PR ?

It is not expanding any functionality besides what already existed in Firecracker main branch.
The guarded pages have already been merged and used for some time. This PR is just removing them in a commit and porting them to the new code in a follow-up commit

sandreim commented on tests/integration_tests/build/test_binary_size.py:
```diff
@@ -13,9 +13,9 @@
+        "FC_BINARY_SIZE_TARGET": 2175176,
         "JAILER_BINARY_SIZE_TARGET": 1439512,
-        "FC_BINARY_SIZE_LIMIT": 2171516,
+        "FC_BINARY_SIZE_LIMIT": 2177024,
```
This is interesting - binary getting larger while not adding additional functionality.

alindima replied:
I was suprised by this also. It's due to this issue: https://github.com/firecracker-microvm/firecracker/issues/2699

sandreim reviewed (commented):
Changes look good overall. How far are we from dropping the local `vm-memory` crate from FC ?

dianpopa commented on src/vm-memory/src/lib.rs:
```diff
@@ -5,18 +5,355 @@
+const GUARD_NUMBER: usize = 2;
+
+/// Build a MmapRegion surrounded by guard pages.
+fn build_guarded_region(
```
have we  ever considered in the past moving this function to upstream? Building guarded memory regions is something other vmm could benefit from

alindima replied:
see this comment: https://github.com/firecracker-microvm/firecracker/pull/2661#issuecomment-896977110

alindima replied:
> A more detailed description would be helpful here - hinting the nested mmap() layout.

Good callout, will do this

dianpopa commented on src/vm-memory/Cargo.toml:
```diff
@@ -5,6 +5,6 @@ authors = ["Amazon Firecracker team <firecracker-devel@amazon.com>"]
 
 [dependencies]
-vm-memory-upstream = { package = "vm-memory", version = ">=0.2.2", features = ["backend-mmap"] }
+vm-memory-upstream = { package = "vm-memory",  version = "0.6.0", features = ["backend-mmap", "backend-bitmap"] }
```
My take on this is that `create_guest_memory`  should be placed in vmm since this is the glue code needed in order to make use of upstream vm-memory so i do not see the need to have a separate crate in fc for this

alindima replied:
see this comment: https://github.com/firecracker-microvm/firecracker/pull/2661#issuecomment-896977110

alindima commented on src/vm-memory/src/lib.rs:
```diff
@@ -5,18 +5,117 @@
+    flags: i32,
+    track_dirty_pages: bool,
+) -> Result<MmapRegion, MmapRegionError> {
+    let page_size = unsafe { libc::sysconf(libc::_SC_PAGESIZE) as usize };
```
> Simplifying assumption: can we assume page size 4k, and not even call sysconf ?

I see that we assume 4K pages in multiple places across our codebase. Is this documented anywhere?
However, I think it's best to have a safe helper in `utils` for checking this, as @serban300 mentioned

alindima commented on src/arch/src/aarch64/fdt.rs:
```diff
@@ -660,7 +660,14 @@ mod tests {
     fn test_create_fdt_with_devices() {
         let regions = arch_memory_regions(layout::FDT_MAX_SIZE + 0x1000);
-        let mem = GuestMemoryMmap::from_ranges(&regions).expect("Cannot initialize memory");
+        let mem = vm_memory::create_guest_memory(
+            &regions
+                .into_iter()
+                .map(|r| (None, r.0, r.1))
+                .collect::<Vec<_>>(),
+            false,
+        )
+        .expect("Cannot initialize memory");
```
The only difference for this test helper would be that it had a parameter `regions: &[(GuestAddress, usize)]` instead of `regions: &[(Option<FileOffset>, GuestAddress, usize)]`.
I cannot think of a good name for such a function but if you have a good suggestion, please do tell.
I also think that for such a small argument difference, a separate function may be overkill, WDYT?

serban300 replied:
What about `create_anonymous_guest_memory()` ? Since it's not file-backed. I think it would make sense having this separate method since now we duplicate the following code segment quite a lot.

```
            &regions
                .into_iter()
                .map(|r| (None, r.0, r.1))
                .collect::<Vec<_>>(),
```

alindima replied:
sounds fair, will do

alindima replied:
done

alindima commented on src/vm-memory/src/lib.rs:
```diff
@@ -155,3 +155,205 @@ pub mod test_utils {
+        let page_size = unsafe { libc::sysconf(libc::_SC_PAGESIZE) as usize };
+
+        let pid = unsafe { libc::fork() };
+        match pid {
+            0 => {
+                let mut addr = region.as_ptr() as usize;
+                let right_page_guard = addr + region.size();
+                loop {
+                    unsafe {
+                        std::ptr::write(addr as *mut u8, 0xFF);
+                    }
+
+                    if addr == right_page_guard {
+                        break;
+                    }
+                    addr += page_size;
+                }
+                unsafe {
+                    libc::exit(0);
+                }
+            }
+            child_pid => {
+                let mut child_status: i32 = -1;
+                let pid_done = unsafe { libc::waitpid(child_pid, &mut child_status, 0) };
+                assert_eq!(pid_done, child_pid);
+
+                // Asserts that the child process terminated because
+                // it received a signal that was not handled.
+                assert!(libc::WIFSIGNALED(child_status));
+                // Signal code should be a SIGSEGV (11)
+                assert_eq!(libc::WTERMSIG(child_status), 11);
+            }
```
the logic is indeed pretty similar, but adding a parameter for telling whether it should assert success or sigsegv would not suffice.
This function forks a new process and loops through the entire region.
In contrast, the `apply_operation_on_address()` forks on every call.

Forking for every iteration of the loop should work in theory, but it would require an enormous amount of new processes, and the test would likely timeout.

serban300 replied:
Can't we have the looping logic outside of `apply_operation_on_address()` ? And call `apply_operation_on_address()` just to test if the operation is performed successfully ?

alindima replied:
the problem is the forking logic, since `apply_operation_on_address()` performs a `fork()` before doing the operation (because it may receive sigsegv).
We can take the forking logic outside `apply_operation_on_address()` and create another helper function like `fork_and_apply_operation` that forks and then calls `apply_operation_on_address`, so that we don't have to add the forking logic in the several places that need it

serban300 replied:
Sounds good to me

alindima commented on src/vm-memory/src/lib.rs:
```diff
@@ -155,3 +155,205 @@ pub mod test_utils {
+                match op {
+                    AddrOp::Read => {
+                        let value = unsafe { std::ptr::read(addr) };
+                        // We have to do something with the value, else
+                        // the Release version will optimize it out, making the test fail
+                        println!("SIGSEGV: {}", value);
```
Switched to using [read_volatile](https://doc.rust-lang.org/std/ptr/fn.read_volatile.html)

alindima commented:
> As a high level comment, we usually make sure that each commit builds. I don't know of any exception so far. Can we do the same thing here while keeping the PR structure (keeping it split in multiple small commits in order to make it easy to review) ?

Unfortunately, I can't do this without squashing into the first commit the commits adding modifications to the tests.
I'm inclined to keep the current commits, since they make for easier to review code and better structure.

WDYT @serban300 ?

alindima commented:
We synced offline and decided that keeping the minimal proxy crate in the Firecracker is the right thing to do, for the reasons outlined above.

The PR is now ready for another review, I've addressed all of your comments @serban300 @dianpopa @sandreim

alindima commented:
I've changed the target branch of this PR to be `main`, since vm-memory has had a release that we can consume.

serban300 commented on src/vm-memory/src/lib.rs:
```diff
@@ -5,18 +5,329 @@
+        None => (-1, 0),
+    };
+
+    let region_start_addr = guard_addr as usize + page_size * (GUARD_PAGE_COUNT / 2);
```
This logic could be problematic if we set an odd value for `GUARD_PAGE_COUNT`. Maybe it would be better to have `GUARD_PAGE_COUNT_BEFORE`, `GUARD_PAGE_COUNT_AFTER` in order to make everything clear. Or we could have just `GUARD_PAGE_COUNT`, but have it represent the number of guard pages for each section (before, after) instead of the total number of guard pages. So in this case `GUARD_PAGE_COUNT` would be `1` and we would change the way we compute `region_start_addr` and `guarded_size`. Would it make sense ?

alindima replied:
good point. I'll fix this

serban300 commented on src/vm-memory/src/lib.rs:
```diff
@@ -46,7 +46,7 @@ fn build_guarded_region(
     track_dirty_pages: bool,
 ) -> Result<MmapRegion, MmapRegionError> {
-    let page_size = unsafe { libc::sysconf(libc::_SC_PAGESIZE) as usize };
+    let page_size = utils::get_page_size().expect("Cannot retrieve page size.");
```
Should we return a `MmapRegionError` instead of panic ?

alindima replied:
I couldn't find any error variants in the vm-memory crate that would hint at this error. I see that it does not even check for this at all: https://github.com/rust-vmm/vm-memory/blob/acd281e9f76a45f331ac1dc3fee87b8ed1044f53/src/mmap_unix.rs#L199
I also think that it should never happen in practice, so I propose we leave it like this for now.

alindima commented:
I am noticing some block throughput degradation with this PR, so let's wait until we conclude the investigation

serban300 commented:
> I am noticing some block throughput degradation with this PR, so let's wait until we conclude the investigation

This reminds me of an issue. Do you think we could try to address https://github.com/firecracker-microvm/firecracker/issues/2592 in this PR as well ? Or right after ? Should be easy to implement and it would confirm if the new vm-memory release offers the expected performance for `read_from` and `write_to`.

alindima commented:
> > I am noticing some block throughput degradation with this PR, so let's wait until we conclude the investigation
> 
> This reminds me of an issue. Do you think we could try to address #2592 in this PR as well ? Or right after ? Should be easy to implement and it would confirm if the new vm-memory release offers the expected performance for `read_from` and `write_to`.

good call! This PR does in fact unblock the performance side of the issue. The error handling is not solved so I posted an update and opened an issue in rust-vmm: https://github.com/rust-vmm/vm-memory/issues/171

alindima commented:
> I am noticing some block throughput degradation with this PR, so let's wait until we conclude the investigation

Investigation was concluded. No issue is caused by vm-memory, so we are unblocking this PR
